ls1intum/Artemis
# `Development`: Modernize appearance of charts on programming exercise grading page

### Checklist
#### General


#### Client

### Motivation and Context


This PR is changing the underlying framework for the charts located in programming exercise grading from `ng2-charts`/`chart.js `to `ngx-charts`.

### Description

The html templates as well as the underlying logic in the components are refactored in order to fit the requests of `ngx-charts`.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 student
- 1 Programming Exercise with static code analysis enabled
- It might be easier to have two windows with the users in parallel rather than logging out and logging in again. This can be realized via different browsers or using incognito mode.

**Important**: Before deploying this PR, take a look at the current charts. Especially check the tooltips and that the colors in the list on the left and in the charts are mapped correctly (e.g. light blue bar represents value of the list entry associated with the color light blue). Taking screenshots might ease the comparison of the different designs.

1. Log in to Artemis as instructor
2. Navigate to Course Administration
3. Click on "Exercises" of the corresponding course
4. In the row of the programming exercise click on the "Grading" button
5. Navigate to the tab "Code Analysis"
6. In the list on the right, feel free to play with the configuration parameters. Inspect if the chart still changes as with the old implementation. (afaik with no submissions, only the first bar should change)
7. **Don't forget to save your changes**
8. Go back to the "Test cases" tab. Again, be creative with the different parameters and make sure that the Weight distribution chart still works as before. ("Total points" should stay empty if nobody participated on the exercise)
9.  **Don't forget to save your changes**
10. Switch to the student account **but if possible keep the grading page open for the instructor**
11. Participate on the exercise and submit your participation. Try to trigger the static code analysis e.g. by inserting unused public attributes.
12. After the submission is build, as instructor switch to the grading page again and go to the "Code Analysis" tab (maybe refreshing might be necessary but not sure). Make sure that the chart is updated correctly.
13. Go back to the "Test Cases" tab. If your submission achieved any points, "Total points" should be updated appropriately. 
14. For all of those charts, the tooltips  should be translated now (which was not the case before). Check this as well.
### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage






### Screenshots


Old design of Category Distribution:
![image](https://user-images.githubusercontent.com/80622272/145387191-24494b8b-2d36-4fbc-aa41-5faa56e7e0a2.png)

*New* Design of Category Distribution:
![image](https://user-images.githubusercontent.com/80622272/145386467-69de0eb0-fabc-4fb2-8b79-56ef3df1d5ee.png)

Old Design of Weight Distribution:
![image](https://user-images.githubusercontent.com/80622272/145387334-fc12f0b9-8e2a-49be-8c16-e9ddf2978df8.png)

*New* Design of Weight Distribution:
![image](https://user-images.githubusercontent.com/80622272/145386629-7ef7bf86-d7ba-48a8-9c82-e7ce537be6ce.png)

Old Design of Total Points:
![image](https://user-images.githubusercontent.com/80622272/145387429-d49149cb-e312-4a9c-825b-8f80a08cd905.png)

*New* Design of Total Points:
![image](https://user-images.githubusercontent.com/80622272/145386729-a0a36d06-c597-4e82-811e-796db0edb9e4.png)
